redis/src/sentinel.rs

```diff
@@ -0,0 +1,558 @@
+//! Defines a Sentinel type that connects to Redis sentinels and creates clients to
+//! master or replica nodes.
+//!
+//! # Example
+//! ```rust,no_run
+//! use redis::Commands;
+//! use redis::sentinel::Sentinel;
+//!
+//! let nodes = vec!["redis://127.0.0.1:6379/", "redis://127.0.0.1:6378/", "redis://127.0.0.1:6377/"];
+//! let sentinel = Sentinel::build(nodes).unwrap();
+//! let mut master = sentinel.master_for("master_name").unwrap().get_connection().unwrap();
+//! let mut replica = sentinel.replica_for("master_name").unwrap().get_connection().unwrap();
+//!
+//! let _: () = master.set("test", "test_data").unwrap();
+//! let rv: String = replica.get("test").unwrap();
+//!
+//! assert_eq!(rv, "test_data");
+//! ```
+//!
+//! There is also a SentinelClient which acts like a regular Client, providing the
+//! `get_connection` and `get_async_connection` methods, internally using the Sentinel
+//! type to create clients on demand for the desired node type (Master or Replica).
+//!
+//! # Example
+//! ```rust,no_run
+//! use redis::Commands;
+//! use redis::sentinel::{ SentinelServerType, SentinelClient };
+//!
+//! let nodes = vec!["redis://127.0.0.1:6379/", "redis://127.0.0.1:6378/", "redis://127.0.0.1:6377/"];
+//! let master_client = SentinelClient::build(nodes, "master_name", SentinelServerType::Master).unwrap();
+//! let replica_client = SentinelClient::build(nodes, "master_name", SentinelServerType::Replica).unwrap();
+//! let mut master_conn = master_client.get_connection().unwrap();
+//! let mut replica_conn = replica_client.get_connection().unwrap();
+//!
+//! let _: () = master_conn.set("test", "test_data").unwrap();
+//! let rv: String = replica_conn.get("test").unwrap();
+//!
+//! assert_eq!(rv, "test_data");
+//! ```
+//!
+
+use std::{
+    collections::HashMap,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+#[cfg(feature = "aio")]
+use std::future::Future;
+
+#[cfg(feature = "aio")]
+use futures_util::StreamExt;
+
+use crate::{
+    connection::ConnectionInfo, types::RedisResult, Client, Connection, ErrorKind, FromRedisValue,
+    IntoConnectionInfo, Value,
+};
+
+/// The Sentinel type, serves as a special purpose client which builds other clients on
+/// demand.
+#[derive(Debug, Clone)]
+pub struct Sentinel {
+    sentinels_connection_info: Vec<ConnectionInfo>,
+    replica_start_index: usize,
+}
+
+fn sentinel_masters_cmd() -> crate::Cmd {
+    let mut cmd = crate::cmd("SENTINEL");
+    cmd.arg("MASTERS");
+    cmd
+}
+
+fn sentinel_replicas_cmd(master_name: &str) -> crate::Cmd {
+    let mut cmd = crate::cmd("SENTINEL");
+    cmd.arg("SLAVES"); // For compatibility with older redis versions
+    cmd.arg(master_name);
+    cmd
+}
+
+#[cfg(feature = "aio")]
+async fn async_try_sentinel_masters(
+    connection_info: &ConnectionInfo,
+) -> RedisResult<Vec<HashMap<String, String>>> {
```

Comment:
This is because I'm using the `SENTINEL MASTERS` command, which may return multiple masters, instead of `SENTINEL get-master-addr-by-name master-name`. The latter is actually the suggested command in the sentinel client doc, but I the former is what redis-py uses, and I believe it's because only this command returns the master's flags (which are used to identify whether a master is valid).

The return of this method is then used by `async_get_first_valid_master` to find the master with the desired name (and now I'm thinking that this method name doesn´t make much sense, since there can be only one master with each name).

Reply:
OK, so this isn't intuitive. Please document the functions, to clarify the intended flow. alternatively, call `async_get_first_valid_master ` from this function, and return the single master.

Reply:
I tried adding some comments and renaming the `async_get_first_valid_master` function. I was not sure if I should repeat the comments for the async functions, so for now I just reference the non-async function.